File: Fall Week 6 Graphs/c++/solutions/party.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;


int main()
{
    int n;
    cin >> n;
    vector<int> manager(n);
    for(int i = 0; i < n; i++) {
        cin >> manager[i];
        if(manager[i] != -1) {
            manager[i] -= 1; // convert 1-index to 0-index
        }
    }

    vector<vector<int>> adj(n);
    for(int i = 0; i < n; i++) {
        int j = manager[i];
        if(j == -1) continue;
        // we make an edge from the manager to the employee
        adj[j].push_back(i);
    }
    // our goal is to find the length of the manager chain to each person
    vector<int> dists(n, -1); // use -1 to mean distance hasn't been found yet
    for(int i = 0; i < n; i++) {
        if(manager[i] != -1) continue;
        // we start looking for the length of the manager chain using bfs
        queue<int> bfs;
        bfs.push(i);
        dists[i] = 1; // chain starts with length 1 (just the manager)
        while(!bfs.empty()) {
            int manager = bfs.front(); bfs.pop(); // get front of queue and remove it
            for(int employee: adj[manager]) {
                // loop through the vector

                /*
                Note: we don't need to worry about whether a node has been visited cause
                each one can only be visited once

                Why? There is only one manager for each employee
                */
                
                dists[employee] = dists[manager] + 1;
                bfs.push(employee);
            }
        }
    }

    int longestChain = 0;
    for(int i = 0; i < n; i++) {
        if(dists[i] > longestChain) {
            longestChain = dists[i];
        }
    }

    cout << longestChain << '\n';
}

File: Fall Week 6 Graphs/python/solutions/party.py
from collections import deque

n = int(input())
manager = [0] * n
for i in range(n):
    manager[i] = int(input())
    if manager[i] != -1:
        manager[i] -= 1  # convert 1-index to 0-index

adj = [[] for _ in range(n)] # constructing 2-d lists is tricky in Python
for i in range(n):
    j = manager[i]
    if j == -1:
        continue
    # we make an edge from the manager to the employee
    adj[j].append(i)

# our goal is to find the length of the manager chain to each person
dists = [-1] * n  # use -1 to mean distance hasn't been found yet
for i in range(n):
    if manager[i] != -1:
        continue
    # we start looking for the length of the manager chain using bfs
    bfs = deque() # python doesn't have a queue, so use deque
    bfs.append(i)
    dists[i] = 1  # chain starts with length 1 (just the manager)
    while bfs:
        cur = bfs.popleft()  # get front of queue and remove it
        for employee in adj[cur]:
            # loop through the vector

            '''
            Note: we don't need to worry about whether a node has been visited cause
            each one can only be visited once

            Why? There is only one manager for each employee
            '''

            dists[employee] = dists[cur] + 1
            bfs.append(employee)

longest_chain = 0
for i in range(n):
    if dists[i] > longest_chain:
        longest_chain = dists[i]

print(longest_chain)
